src/main/java/com/google/sps/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,266 @@
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Handles POST and GET requests for projects,
+ * allows creation and update of projects, and
+ * supports queries for projects based on various parameters.
+ */
+@WebServlet("/projects")
+public class ProjectServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Must be logged in
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/"); // placeholder: should redirect to login
+      return;
+    }
+
+    // Mode is a required parameter
+    boolean isCreateMode = DataUtils.parseMode(request, response);
+
+    String uEmail = userService.getCurrentUser().getEmail();
+    String projId = request.getParameter("proj-id");
+
+    // Will be either a new entity for creation or an existing entity for
+    // updating
+    Entity projEntity = new Entity(DataUtils.PROJECT);
+
+    if (!isCreateMode) {
+      // Must be owner to update
+      projEntity = DataUtils.getProjectEntity(projId, uEmail, false, false);
+
+      // Delete overrides all other updates
+      boolean delete = Boolean.parseBoolean(request.getParameter("delete"));
+      if (delete) {
+        datastore.delete(projEntity.getKey());
+        response.sendRedirect("/"); // placeholder: should redirect to projects gallery
+        return;
+      }
+    }
+
+    // Last modified timestamp
+    String now = Instant.now().toString();
+    projEntity.setProperty("utc", now);
+
+    String projName = request.getParameter("proj-name");
+
+    // Set the project name if provided
+    // If creating and nothing provided, set name to Untitled-{current UTC time}
+    if (!DataUtils.isEmptyParameter(projName)) {
+      projEntity.setProperty("name", projName);
+    } else {
+      if (isCreateMode) {
+        projEntity.setProperty("name", "Untitled-" + now);
+      }
+      // Don't do anything if updating and no name provided
+    }
+
+    // Default visibility to private (only owners and editors can view)
+    String visibility = request.getParameter("visibility");
+    if (DataUtils.isEmptyParameter(visibility) && isCreateMode) {
+      visibility = DataUtils.PRIVATE;
+    }
+    if (!DataUtils.isEmptyParameter(visibility) &&
```

Comment:
else if?

Reply:
Both statements may be run, which is why there are two if statements. The first one handles the default visibility for create mode and the second one actually sets the entity property.